### PR TITLE
fix: rollback should not be invoked on steps that never executed

### PR DESCRIPTION
**Pull request by leninmehedy**
## Description

This pull request updates the workflow rollback logic to ensure that rollback is only performed for steps that actually executed, in compliance with spec D5 (§5.3). Steps that failed during the Prepare phase are now excluded from rollback, preventing their compensating actions from running. The PR also adds tracking for executed steps and updates tests to cover this behavior.

**Rollback logic improvements:**

* Added a `lastExecutedStepIDs` field to the `workflow` struct to track which steps actually reached `Execute`.
* Modified the `rollbackFrom` method to accept an `executedIDs` parameter and skip rollback for steps that did not execute, marking their rollback reports as `Skipped`.
* Updated the `Execute` method to track executed steps and pass them to `rollbackFrom` during rollback, ensuring only executed steps are compensated [[1]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafR389-R391) [[2]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafR465) [[3]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafL508-R534) [[4]](diffhunk://#diff-f789a7245fb48ae3d30877bd241f7a279a2eec3b4de705284a3e39d3f0cd9dafL522-R553).
* Updated the `Rollback` method to use the new executed-step tracking when performing a manual rollback.

**Testing:**

* Added the test `TestWorkflow_RollbackSkipsPrepareFailed` to verify that steps failing in `Prepare` do not have their `Rollback` called, and their rollback report status is `Skipped`.
* Updated existing tests to use the new `rollbackFrom` signature [[1]](diffhunk://#diff-69a2723318db0a4416df092fc13f68f8e05454975c93b361dd777d7b19eb0e97L225-R225) [[2]](diffhunk://#diff-69a2723318db0a4416df092fc13f68f8e05454975c93b361dd777d7b19eb0e97L251-R251) [[3]](diffhunk://#diff-69a2723318db0a4416df092fc13f68f8e05454975c93b361dd777d7b19eb0e97L277-R277) [[4]](diffhunk://#diff-69a2723318db0a4416df092fc13f68f8e05454975c93b361dd777d7b19eb0e97L303-R303).

### Related Issues

* Closes #
* Related to #

### Pull request (PR) checklist

* \[ ] This PR added tests (unit, integration, and/or end-to-end)
* \[ ] This PR updated documentation
* \[ ] This PR added no TODOs or commented out code
* \[ ] This PR has no breaking changes
* \[ ] Any technical debt has been documented as a separate issue and linked to this PR
* \[ ] Any `go.mod` and/or `go.sum` changes have been explained to and approved by a repository manager
* \[ ] All related issues have been linked to this PR
* \[ ] All changes in this PR are included in the description
* \[ ] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit
  message guidelines' below have been followed

### Testing

* \[ ] This PR added unit tests
* \[ ] This PR added integration/end-to-end tests
* \[ ] These changes required manual testing that is documented below
* \[ ] Anything not tested is documented

The following manual testing was done:

* TBD

The following was not tested:

* TBD

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a
   breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any
   type. NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                       |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

File: workflow.go [hunk 1]
```diff
@@ -519,12 +545,12 @@ func (w *workflow) Execute(ctx context.Context) *Report {
...
 
-	// Preserve state snapshots for potential manual rollback later
-	// (even if workflow failed, manual Rollback() can use these snapshots)
-	// Only preserve if configured to reduce memory overhead when rollback isn't needed
+	// Preserve state snapshots and executed-step set for potential manual rollback later.
+	// Only preserve states if configured to reduce memory overhead when rollback isn't needed.
 	if w.preserveStatesForRollback {
 		w.lastExecutionStates = stepStates
 	}
+	w.lastExecutedStepIDs = executedStepIDs
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow.go:555]: Fixed. `lastExecutionStates` is now explicitly set to `nil` in the `else` branch so stale snapshots from a prior execution with preservation enabled are always cleared when preservation is off.

File: workflow.go [hunk 2]
```diff
@@ -617,8 +643,8 @@ func (w *workflow) Rollback(ctx context.Context) *Report {
 
 	w.log().Info("starting workflow rollback", "workflowId", w.id, "steps", len(w.steps))
 
-	// Use preserved states from last execution, fall back to nil if none available
-	rollbackReports := w.rollbackFrom(ctx, len(w.steps)-1, w.lastExecutionStates)
+	// Use preserved states and executed-step set from last execution.
+	rollbackReports := w.rollbackFrom(ctx, len(w.steps)-1, w.lastExecutionStates, w.lastExecutedStepIDs)
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow.go:649]: Fixed. `rollbackFrom` now treats `nil` as empty (not as "no filter") — a nil map lookup always returns the zero value and `!ran` is true, so every step is skipped. This means hand-constructed workflows with a nil `lastExecutedStepIDs` correctly compensate nothing. `newDefaultWorkflow` and `newTestWorkflow` also initialize the field to an empty map explicitly for clarity, and the direct `rollbackFrom` tests now pass an explicit executed-set rather than nil.
